### ISSUE TITLE
Regression in desktop switcher not showing the correct desktop

### ISSUE THREAD

**Issue by ssipos90** [lxqt/lxqt-panel#2173]
##### Expected Behavior

The desktop switcher should reflect the current desktop.

##### Current Behavior

Since upgrading to version 2.1.1, the desktop switcher is wonky and doesn't reflect the current desktop in all situations, especially when moving away from an empty desktop. It seems that something's off with the dynamic desktops detection that i3wm uses.

Made a short video using screenkey showing the behaviour:
https://github.com/user-attachments/assets/55860edc-aabd-4985-9490-1299b40f32b9

##### Possible Solution
This is a regression, as it worked up until and including 2.0.1. 

##### Steps to Reproduce (for bugs)
I think it's enough to have i3wm declare more than 5 desktops.

##### System Information
* WM: i3wm
* LXQt Version: 2.1.0
* LXQt Panel version: 2.1.1
* Distribution & Version: Manjaro up to date
* Qt Version: 5.15.15
* liblxqt Version: 2.1.0-1

**Comment by stefonarch** [comment]
I never used i3, only sway sometimes. But I think I can confirm this issue. I see that I can only create one empty workspace. If I open an app on 1, another on 2 and then create 3 and open an app there  and close the one on 2 when cycling the workspaces  the switcher show only 1 2 and not 1 2 3.

There was a refactoring needed for wayland plugins, probably something has changed, other WM are fine.
@gfgit any idea?

**Comment by tsujan** [comment]
> But I think I can confirm this issue

What about other X11 WM? Do they show this problem?

**Comment by tsujan** [comment]
> other WM are fine.

But we don't treat WMs separately; we use KF6 for all. How can this be a new issue?

**Comment by stefonarch** [comment]
Afaik no other WM uses dinamic workspaces. I can't see this issue in openbox or xfwm4.
@ssipos90 How can I declare 5 workspaces?

**Comment by tsujan** [comment]
What I wanted to know is just this: Did our switcher worked fine with *dynamic workspaces* before? If the answer is "yes", it gives motivation (to me, for example) for checking whether there's a real difference between our new and old X11 codes, because the last time I checked, there wasn't.

**Comment by tsujan** [comment]
After a quick comparison, I found a regression:  the function `DesktopSwitch::onDesktopNamesChanged()` is never called, because we missed a connection to `KX11Extras::desktopNamesChanged` in `lxqtwmbackend_x11.cpp`.

I'll add that connection tomorrow, but it's about changing the desktop names, and I don't know if it can have any relation to this issue.

**Comment by tsujan** [comment]
> I'm moving from desktop 1 (browser) to 2 (empty) then to 3 (terminal), but it doesn't reflect desktop 3.

Your video shows a different thing:

 * There were 5 desktops at first.
 * Then 2 *was created* after 1, such that the number of desktops became 6.
 * Then the user switched to 3 and *2 was removed*, such that now the number of desktops became 5 again, with 3 as the second desktop, and so, being shown as 2.
 * At the end of the video, 2 was created again and the number of desktops became 6.

There's nothing wrong.

@stefonarch 
Your description doesn't show any problem either. You said, "omitting the empty 2 and 3 will be shown as 2". Well, it should because now it's the second desktop.

**Comment by ssipos90** [comment]
I'll install the previous version and record a short video with the previous behaviour.

**Comment by ssipos90** [comment]
> Your video shows a different thing:
> 
>     There were 5 desktops at first.

@tsujan unfortunately, no. The new panel version thinks that, but it's not the case as i3 deletes empty workspaces.

**Comment by ssipos90** [comment]
Here is what `$ xprop -root` returns, compared to the panel toolbar.

![lxqt-panel-3](https://github.com/user-attachments/assets/e9264e35-3b4a-4d69-aa32-f865760e1343)

**Comment by tsujan** [comment]
If i3wm explicitly sets the name of the second desktop to "3" after "2" is removed, then https://github.com/lxqt/lxqt-panel/pull/2174 should fix this.

**Comment by ssipos90** [comment]
> I'll install the previous version and record a short video with the previous behaviour.
> 
> edit:
> 
> **(desktop switcher is set to names, not numbers)**

I mentioned I'm using names, not numbers. It's not that.

**Comment by tsujan** [comment]
As I said, I tested it (with https://github.com/lxqt/lxqt-panel/pull/2174), and it was OK.

**Comment by ssipos90** [comment]
Ah ok. I didn't properly follow the latest comments.

I just cloned and build from source from that branch and it works great.

Thank you!!

**Comment by tsujan** [comment]
Then I also misunderstood it, in a way.

You mean that only after using https://github.com/lxqt/lxqt-panel/pull/2174 the problem was fixed?

**Comment by ssipos90** [comment]
yes. v2.1.1 doesn't work properly. I always, since day 1 had "names" set since I use icons.

**Comment by tsujan** [comment]
OK. So, actually fixing of the missing signal `workspaceNameChanged` did it, although I didn't hope that it would be relevant to this report.

Thanks for reporting and testing!

**Comment by ssipos90** [comment]
you too for fixing!

**Comment by tsujan** [comment]
@stefonarch 
Do you think this needs a point release? If you say "yes", I'll make a point release for `lxqt-wayland-session` too — you know why ;)

**Comment by stefonarch** [comment]
Yes :)

**Comment by tsujan** [comment]
:)

Will make them tomorrow.
